axum/src/serve/mod.rs

```diff
@@ -686,4 +701,31 @@ mod tests {
         let body = String::from_utf8(body.to_vec()).unwrap();
         assert_eq!(body, "Hello, World!");
     }
+
+    #[crate::test]
+    async fn serving_with_custom_body_type() {
+        struct CustomBody;
+        impl http_body::Body for CustomBody {
+            type Data = bytes::Bytes;
+            type Error = std::convert::Infallible;
+            fn poll_frame(
+                self: std::pin::Pin<&mut Self>,
+                _cx: &mut std::task::Context<'_>,
+            ) -> std::task::Poll<Option<Result<http_body::Frame<Self::Data>, Self::Error>>>
+            {
+                #![allow(clippy::unreachable)] // The implementation is not used, we just need to provide one.
+                unreachable!();
+            }
+        }
+
+        let app = ServiceBuilder::new()
+            .layer_fn(|_| tower::service_fn(|_| std::future::ready(Ok(Response::new(CustomBody)))))
+            .service(Router::<()>::new());
```

Comment:
If we add some routes, which type will be the body? The axum body type? Or `CustomBody`?

Reply:
Could we add an handle to check if this works? Or are our handlers tight to our body implementation?

Reply:
I can do that (in a few days).

The handler (and `Router`) would still return `axum::body::Body`, it's just the outer layer which changes that to something else.

Reply:
I've added a route there.

Note though that the whole `Router` is completely ignored by the outer layer. I could even write this instead and it would work the same way even without the router.
```
let app = tower::service_fn(|_| std::future::ready(Ok(Response::new(CustomBody))));
```

Reply:
I see. What would be a better example?
A middleware changing the body type depending of the request?
(I have issues understanding the value)

Reply:
#3002 is good example I think. I just wanted to add a test that stuff like this compiles going forward.

We can add a real example or extend the one showing middlewares wrapping the whole router (assuming we already have one)

Reply:
There's this example I wrote: https://github.com/tokio-rs/axum/blob/main/examples/compression/src/main.rs
(but I guess there's no need to layer outside the router for this)
I'll approve this PR code-wise. Feel free to enhance it with examples / tests showing its value as I cannot find any myself.